scrapers/ESP/index.js

```diff
@@ -1,38 +1,128 @@
+import { LocalDate, DateTimeFormatter } from '@js-joda/core';
 import * as fetch from '../../lib/fetch.js';
-import * as parse from '../../lib/parse.js';
 import * as transform from '../../lib/transform.js';
-import * as rules from '../../lib/rules.js';
-
-// Set county to this if you only have state data, but this isn't the entire state
-// const UNASSIGNED = '(unassigned)';
 
 const scraper = {
   country: 'ESP',
-  url: 'https://opendata.arcgis.com/datasets/48fac2d7de0f43f9af938852e3748845_0.csv',
+  url: 'https://github.com/datadista/datasets/tree/master/COVID%2019',
   priority: 1,
+  headless: false,
+  timeseries: true,
   aggregate: 'state',
+  sources: [
+    {
+      name: 'cases',
+      url: 'https://raw.githubusercontent.com/datadista/datasets/master/COVID%2019/ccaa_covid19_casos.csv',
+      description: 'Cases by date and community'
+    },
+    {
+      name: 'recovered',
+      url: 'https://raw.githubusercontent.com/datadista/datasets/master/COVID%2019/ccaa_covid19_altas.csv',
+      description: 'Recoveries by date and community'
+    },
+    {
+      name: 'deaths',
+      url: 'https://raw.githubusercontent.com/datadista/datasets/master/COVID%2019/ccaa_covid19_fallecidos.csv',
+      description: 'Deaths by date and community'
+    }
+  ],
   async scraper() {
-    const data = await fetch.csv(this.url);
-    const states = [];
-    for (const row of data) {
-      const state = row.Texto !== undefined ? parse.string(row.Texto) : '';
-      const cases = row.TotalConfirmados !== undefined ? parse.number(row.TotalConfirmados) : 0;
-      const deaths = row.TotalFallecidos !== undefined ? parse.number(row.TotalFallecidos) : 0;
-      const recovered = row.TotalRecuperados !== undefined ? parse.number(row.TotalRecuperados) : 0;
-      if (state !== '') {
-        const data = {
-          state,
-          cases,
-          deaths,
-          recovered
-        };
-        if (rules.isAcceptable(data, null, this._reject)) {
-          states.push(data);
-        }
-      }
+    const ISO = DateTimeFormatter.ofPattern('yyyy-MM-dd');
+    const isDate = s => s.includes('/');
+
+    const parseDate = s => {
+      // Source data currently uses ddmmyyyy format for dates,
+      // but is (maybe?) planning to switch to yyyymmdd.
+      // See https://github.com/datadista/datasets/issues/14
+      // When/if this happens, replace contents of this function with:
+      // return LocalDate.parse(s)
+
+      const [d, m, y] = s.split('/');
+      return LocalDate.parse(`${y}-${m}-${d}`);
+    };
+
+    const mostRecent = dates => dates.sort().pop();
```

Comment:
Can you take a look at the way we're doing timeseries checks in https://github.com/lazd/coronadatascraper/pull/196 and make sure you handle the same errors?

* Requested date is before timeseries begin (error)
* Requested date is after latest date in timeseries (use latest date)

Reply:
I have some ideas about how we can centralize this logic but don't want to step on anyone's toes if someone's already on it - let me know.